backend/src/main/java/fk/prof/backend/model/policy/ZookeeperBasedPolicyStoreAPI.java

```diff
@@ -0,0 +1,232 @@
+package fk.prof.backend.model.policy;
+
+import com.google.common.io.BaseEncoding;
+import fk.prof.backend.exception.PolicyException;
+import fk.prof.backend.proto.PolicyDTO;
+import fk.prof.backend.util.ZookeeperUtil;
+import io.vertx.core.Future;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.zookeeper.CreateMode;
+import recording.Recorder;
+
+import java.nio.charset.Charset;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Zookeeper based implementation of the policy store
+ * Created by rohit.patiyal on 18/05/17.
+ */
+public class ZookeeperBasedPolicyStoreAPI implements PolicyStoreAPI {
+
+    private static final String POLICY_NODE_PREFIX = "v";
+    private static final String DELIMITER = "/";
+    private final CuratorFramework curatorClient;
+    private final Semaphore setPolicySemaphore = new Semaphore(1);
```

Comment:
Semaphore usually makes sense when multiple threads are permitted to enter a critical section. otherwise a reentrant lock makes the purpose more clearer than a binary semaphore. See implementation in zk backend association store